<a href="https://colab.research.google.com/github/MariaHinestroza04/Analitica-Negocios/blob/main/Reto_2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#**Caso de estudio:**

El presente modelo, utiliza los arboles de decisión como herramienta analitica para identificar a los pacientes quienes tienen un alto riesgo de padecer una enfermedad cardíaca. Puede ser utilizado por entidades prestadoras de servicio de salud como hospitales o aseguradoras. Para llevar a cabo este proceso vamos a analizar las siguientes variables:

**Edad:** Al aumentar la edad se aumenta el riesgo de de sufrir de una enfermedad cardiovascular, por lo tanto es importante tenerlo en cuenta como factor de riesgo biologico.

**Diastolic y Systolic:** Estos factores corresponden a la presión del corazon cuando se contrae y cuando está en reposo, son datos fundamentales para determinar el estado del corazón.

**BMI:** El indice de masa corporal es un factyor que puede ser un factor determinante sobre la salud del paciente sobre todo por la obesidad y salud metabolica del paciente.

0. Se procede con la carga de las librerias

In [ ]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import confusion_matrix

1. Carga de los datos del trabajo

In [ ]:
nxl='/content/sample_data/2. HeartDisease (1).xlsx'
XDB=pd.read_excel(nxl,sheet_name=0)
XD=XDB.iloc[:,[1,2,3,4]]
yd=XDB.iloc[:,0]

2. Implementación del modelo de árboles

In [ ]:
mar=DecisionTreeClassifier(criterion='gini',max_depth=4)
mar.fit(XD, yd)
ydp=mar.predict(XD)

#Se construye la matriz de confusión para analizar el funcionamineto del modelo
cm=confusion_matrix(yd, ydp)
display(cm)
VN=cm[0,0];FP=cm[0,1];FN=cm[1,0];VP=cm[1,1]
#metricas de desempeño
exactitud=((VP+VN)/(VP+VN+FP+FN))*100
print("La exactitud es:",exactitud)
sensibilidad=(VP/(VP+FN))*100
print("La sensibilidad es:",sensibilidad)
especificidad=(VN/(VN+FP))*100
print("La especificidad es:",especificidad)
precisión=(VP/(VP+FP))*100
print("La precisión es:",precisión)

array([[63, 10],
       [20, 87]])

La exactitud es: 83.33333333333334
La sensibilidad es: 81.30841121495327
La especificidad es: 86.3013698630137
La precisión es: 89.69072164948454


3. Despliegue del arbol

In [ ]:
from sklearn.tree import export_graphviz
from pydotplus import graph_from_dot_data
vs=["Edad", "Systolic", "Diastolic", "BMI"]
dot_data=export_graphviz(mar,feature_names=vs)
graph=graph_from_dot_data(dot_data)
graph.write_png("arbol.png")


True

**Analisis de resultados: Matriz de Confusión:**

El modelo de árbol de decisión muestra en forma general un muy buen funcionamiento pues sus resultados superan el 75%. La exactitud del 83.33%, indica una correcta clasificación de los pacientes. En cuanto a la sensibilidad, el modelo logró identificar correctamente el 81.31% de los pacientes que realmente presentan la enfermedad, mostrando una buena capacidad para detectar casos positivos. Por su parte, la especificidad fue del 86.30%, muestra que el modelo reconoce adecuadamente a la mayoría de los pacientes sanos. Finalmente, la precisión del 89.69% indica que, cuando el modelo predice la presencia de enfermedad cardíaca, acierta en casi 9 de 10 casos, lo que refleja un buen desempeño predictivo general.

**Analisis de resultados: Arbol de decisión:**

El árbol de decisión, muestra que la variable más determinante en la clasificación del riesgo de enfermedad cardíaca es la edad, el primer corte de segmento por edad, se identifica que en el grupo mayor a 54.5 años (58 pacientes), 52 presentan enfermedad, lo que representa aproximadamente un 89.6%, evidenciando un nodo muy cercano a 0 (gini = 0.185). Además, en pacientes mayores de 69.5 años se identifica un nodo completamente puro (gini = 0) con 2 de 2 pacientes diagnosticados con enfermedad (100%).

En contraste, en el grupo de pacientes con edad menor o igual a 54.5 años (122 individuos), el comportamiento es más heterogéneo (67 sin enfermedad y 55 con enfermedad), por lo que el modelo requiere variables adicionales como BMI y presión arterial para refinar la clasificación. Por ejemplo, en pacientes jóvenes con presión sistólica mayor a 123  se identifica un nodo puro con 12 de 12 pacientes con enfermedad, lo cual evidencia una regla de alto riesgo. Asimismo, combinaciones específicas de presión diastólica elevada también incrementan significativamente la probabilidad de diagnóstico positivo.

En términos generales, el árbol muestra que el riesgo cardiovascular aumenta mas que todo con la edad, mientras que en pacientes más jóvenes el diagnóstico depende en mayor medida de factores metabólicos y de presión arterial, permitiendo establecer reglas de decisión claras y clínicamente interpretables.

4. Pacientes de la segunda base de datos

In [ ]:
nxl='/content/sample_data/2. HeartDisease (1).xlsx'
XDB1_new = pd.read_excel(nxl, sheet_name=1)
XD_new_patients = XDB1_new.iloc[:, [1, 2, 3, 4]]
ydp_new_correct = mar.predict(XD_new_patients)
display(ydp_new_correct)
XD_new_patients['Prediction'] = ydp_new_correct
display(XD_new_patients)

array([0, 0, 0, 1, 1, 1, 1, 1])

,Age,Systolic,Diastolic,BMI,Prediction
0,33,120,64,33,0
1,55,115,81,32,0
2,51,130,70,20,0
3,65,101,63,18,1
4,62,109,110,38,1
5,60,130,69,35,1
6,61,124,95,36,1
7,53,150,81,41,1
